### ISSUE TITLE
Support Elasticsearch5.0

### ISSUE THREAD

**Issue by martintrojer** [clojurewerkz/elastisch#245]
https://www.elastic.co/guide/en/elasticsearch/reference/5.0/breaking-changes-5.0.html

**Comment by timgluz** [comment]
Hi, 

i'm able to join the forces and i can do some serious contributions next couple of weeks;

**Comment by michaelklishin** [comment]
@timgluz woohoo!

**Comment by nkhadakkar** [comment]
Hi, 
any news on upgrading to 5.0 ?

**Comment by michaelklishin** [comment]
@nkhadakkar no news from me. this is an open source project. You can contribute to it.

**Comment by MerelyAPseudonym** [comment]
*My* priority is to get things polished for a 3.0 stable release (supporting Elasticsearch 2), but after that of course I'd like to see support for Elasticsearch 5, and would enjoy helping to implement that. Unfortunately, things at work have been _extremely_ busy as of late, so I haven’t had the time to help get Elastisch 3 over the finish line.

**Comment by michaelklishin** [comment]
Good point. `3.0` should definitely ship before ElasticSearch 5.0 work starts.

**Comment by jinwg** [comment]
Do we have a pending items list of 3.0 release? just check any help can be offered to start the 5.0 works early..

**Comment by Poogles** [comment]
Started work on 5.0 myself as we've upgraded and had things break.

https://github.com/clojurewerkz/elastisch/pull/247

**Comment by neverfox** [comment]
It's also worth pointing out that 1.7 hits EOL on January 16 and Elastic Cloud is pulling the plug on 1.x installations on July 17 through forced upgrades for customer clusters to 2.x. If this library doesn't at least officially support 2.x, Clojure is going to be without any real Elasticsearch client as far as Elastic-supported versions go. The clock is ticking, as they say. Depending on how close the master branch is to full 2.x support, is worth shooting for that before thinking about 5? Or is there a belief that 5 support can come in time?

**Comment by michaelklishin** [comment]
`2.x` has been supported in master and preview releases for a long time. There will be no jump straight to 5.0 because ElasticSearch 2.x installations aren't going away with the 5.0 release regardless of what Elastic supports.

**Comment by neverfox** [comment]
@michaelklishin So is the README incorrect when it says "Elastisch master provides Elasticsearch 2.x compatibility but not every Elasticsearch 2.3+ feature"? I guess what I'm saying is that currently, 2.x is being treated as beta functionality, in terms of how it has been released and documented. If that's no longer then case, then there should be a non-beta release so that 2.x support is clearly established. If it is still the case, I'm suggesting that contributors might want to consider getting it there first. Also, sorry if I wasn't clear, but I wasn't suggesting that 2.x was going away, only 1.x.

**Comment by michaelklishin** [comment]
@neverfox there will be, when the last few issues for that milestone are reviewed and an RC is out for a week or two. Supporting every ElasticSearch feature is no longer a requirement: there are too many of them and their actual use follows the Pareto principle. Please take this discussion to the mailing list as it has little to do with 5.0 beyond what was already said.

**Comment by coreasync** [comment]
I wonder if there is a chance to get released v.3.0.0-RC1. I reviewed all commits since v.3.0.0-beta1 and they seem to be an improvement for ppl wanting to use ES 2.4.

**Comment by michaelklishin** [comment]
@coreasync I'd like to address #192 first.

**Comment by michaelklishin** [comment]
3.0 is out and I'm happy to begin working on this (to the extent I have the time, which admittedly is not a lot).

**Comment by briantrice** [comment]
@michaelklishin I'm strongly motivated to contribute to 5.0 support, as @hantuzun also indicated on https://github.com/clojurewerkz/elastisch/issues/257. Can a branch be started or PRs ready to accept? I'm preparing the `content-type` fix for PR shortly.
